### PR TITLE
feat: add sensor selection from list in configuration via UI

### PR DESCRIPTION
**Pull request by IATkachenko**
While configuration it is comfortable just to click at sensor.
Allow user doing this.

Not all sensors may be in device registry, but we can't check sensors out of registry. We should not limit advanced users in their choose (just because if root tries `rm -rf /`, it should be root problem, not OS), but should help ordinary users.

Alternative version of #105

### CODE REVIEW

**Comment by IATkachenko** [comment]
Will squash/rebase commits after finishing discussion.

**Comment by rautesamtr** [comment]
```bash
  File "/home/rautesamtr/Documents/home_assistant/config/custom_components/thermal_comfort/config_flow.py", line 48, in get_sensors_by_device_class
    - {e.entity_id for e in _er.entities}
  File "/home/rautesamtr/Documents/home_assistant/config/custom_components/thermal_comfort/config_flow.py", line 48, in <setcomp>
    - {e.entity_id for e in _er.entities}
AttributeError: 'str' object has no attribute 'entity_id'
```

**Comment by IATkachenko** [comment]
Sorry, was out of HA to test. Fixed now.

File: custom_components/thermal_comfort/config_flow.py [hunk 1]
```diff
@@ -21,6 +27,28 @@
...
+def get_sensors_by_device_class(
+    _er: EntityRegistry,
+    _hass: HomeAssistant,
+    device_class: str,
+    include_not_in_registry: bool = False,
+) -> list:
+    """Get sensors of required class from entity registry."""
+
+    result = []
+    for d in _er.async_get_device_class_lookup(
```
Comment (rautesamtr) [inline custom_components/thermal_comfort/config_flow.py:39]: I'd prefer if we stick to more pythonic variable names

File: custom_components/thermal_comfort/config_flow.py [hunk 2]
```diff
@@ -21,6 +27,28 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+def get_sensors_by_device_class(
+    _er: EntityRegistry,
```
Comment (rautesamtr) [inline custom_components/thermal_comfort/config_flow.py:31]: I'd prefer if we stick to more pythonic variable names

File: custom_components/thermal_comfort/config_flow.py [hunk 3]
```diff
@@ -21,6 +27,28 @@
...
+    include_not_in_registry: bool = False,
+) -> list:
+    """Get sensors of required class from entity registry."""
+
+    result = []
+    for d in _er.async_get_device_class_lookup(
+        {(Platform.SENSOR, device_class)}
+    ).values():
+        for e in d.values():
+            result.append(e)
```
Comment (rautesamtr) [inline custom_components/thermal_comfort/config_flow.py:43]:
```python
    result = list(
        _er.async_get_device_class_lookup({(Platform.SENSOR, device_class)}).values()
    )
```
No need for a for loop here

File: custom_components/thermal_comfort/config_flow.py [hunk 4]
```diff
@@ -21,6 +27,28 @@
...
+
+    if include_not_in_registry:
+        result += list(
+            {e.entity_id for e in _hass.states.async_all()} - set(_er.entities)
+        )
```
Comment (rautesamtr) [inline custom_components/thermal_comfort/config_flow.py:48]: doesn't work as intended, to many sensors missing compared to just
```python
        result += [e.entity_id for e in _hass.states.async_all()]
```

File: custom_components/thermal_comfort/config_flow.py [hunk 5]
```diff
@@ -56,12 +95,12 @@ def build_schema(
             vol.Required(
                 CONF_TEMPERATURE_SENSOR,
                 default=get_value(config_entry, CONF_TEMPERATURE_SENSOR),
-            ): str,
+            ): vol.In(t_sensors),
```
Comment (rautesamtr) [inline custom_components/thermal_comfort/config_flow.py:98]: We should add a default selection e.g. Element 0 so we avoid error messages if a user tries to continue without selection.

**Review by rautesamtr — CHANGES_REQUESTED** [review]
I really like your idea of splitting this into show_advanced. Keep it simple for the average user while still being flexible enough for power users. Great thinking.

I'd prefer if we could avoid using shortened variable names and stick more to the pythonic way.

File: custom_components/thermal_comfort/config_flow.py (hunk 4 above)
Comment (IATkachenko) [reply custom_components/thermal_comfort/config_flow.py:48]: > doesn't work as intended, to many sensors missing compared to just

Yes, many sensor missed. And it is main idea: all sensors from registry have device class, so this add only out of device class sensors. It allow us to skip pressure sensors from list of temperature, for example.

Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:48]: Yes but there are also temperature sensors completely missing that make no sense not to be available in the list.
e.g. this one
image

File: custom_components/thermal_comfort/config_flow.py (hunk 3 above)
Comment (IATkachenko) [reply custom_components/thermal_comfort/config_flow.py:43]: `async_get_device_class_lookup` returns `dict[str, dict[tuple[Any, Any], str]]` and entity_id is inside second dict. So double loop.

File: custom_components/thermal_comfort/config_flow.py (hunk 4 above)
Comment (IATkachenko) [reply custom_components/thermal_comfort/config_flow.py:48]: > Yes but there are also temperature sensors completely missing that make no sense not to be available in the list. e.g. this one image

Is this sensor in device registry, but have no device_class `DEVICE_CLASS_TEMPERATURE`?

Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:48]: You can get that one with:
```yaml
sensor:
  - platform: demo
```

**Comment by IATkachenko** [comment]
> I'd prefer if we could avoid using shortened variable names and stick more to the pythonic way.

I'm using one letter variables only as loop variable. `_er` -- yes, should be renamed. 
Will use more common names everywhere.

File: custom_components/thermal_comfort/config_flow.py (hunk 3 above)
Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:43]: ah sorry my bad, we could still go for list comprehension for performance
```python
    result = [
        entity.values()
        for entity in _er.async_get_device_class_lookup(
            {(Platform.SENSOR, device_class)}
        ).values()
    ]
```

File: custom_components/thermal_comfort/config_flow.py (hunk 4 above)
Comment (IATkachenko) [reply custom_components/thermal_comfort/config_flow.py:48]: Will test this setup.

Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:48]:
```json
{
   "entity_id=""sensor.outside_temperature",
   "unique_id=""sensor_1",
   "platform=""demo",
   "area_id=""living_room",
   "capabilities="{
      "state_class":"measurement"
   },
   "config_entry_id=None",
   "device_class=""temperature",
   "device_id=None",
   "disabled_by=None",
   "entity_category=None",
   "icon=None",
   "id=""ab12532e0a3e6f14541e131efb79c1ce",
   "name=None",
   "options="{
      
   },
   "original_device_class=""temperature",
   "original_icon=None",
   "original_name=""Outside Temperature",
   supported_features=0,
   "unit_of_measurement=""°C"
}
```
It does have device class temperature and is in the registry but isn't part of `result`.

Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:48]: I also get an empty result to begin with

File: custom_components/thermal_comfort/config_flow.py [hunk 6]
```diff
@@ -4,9 +4,15 @@
 import logging
 
 from homeassistant import config_entries
-from homeassistant.const import CONF_NAME
+from homeassistant.const import (
+    CONF_NAME,
+    DEVICE_CLASS_HUMIDITY,
+    DEVICE_CLASS_TEMPERATURE,
```
Comment (rautesamtr) [inline custom_components/thermal_comfort/config_flow.py:10]: we should use SensorDeviceClass

Comment (IATkachenko) [reply custom_components/thermal_comfort/config_flow.py:10]: I'll check more careful device classes.

Current version works well for my installation, but looks like not for any.
Inside `async_get_device_class_lookup` it do
```python
            device_class = entity.device_class or entity.original_device_class
            domain_device_class = (entity.domain, device_class)
```
so it looks like natural device class should be used.

Will review device filtering.

Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:10]: [# DEVICE_CLASS_* below are deprecated as of 2021.12](https://github.com/home-assistant/core/blob/e92078cf507089908d103b358acf12230ec4312d/homeassistant/const.py#L274)

Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:10]: https://github.com/home-assistant/core/blob/e92078cf507089908d103b358acf12230ec4312d/homeassistant/const.py#L274

File: custom_components/thermal_comfort/config_flow.py (hunk 4 above)
Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:48]: filtering works for me

```python
    result = [
        entity.entity_id
        for entity in filter(
            lambda entity: entity.device_class == device_class
            and entity.domain == Platform.SENSOR,
            _er.entities.values(),
        )
    ]
```

Comment (rautesamtr) [reply custom_components/thermal_comfort/config_flow.py:48]: but helpers are still missing.

File: custom_components/thermal_comfort/config_flow.py (hunk 6 above)
Comment (IATkachenko) [reply custom_components/thermal_comfort/config_flow.py:10]: Oh... Missed deprecation info...

**Comment by rautesamtr** [comment]
My suggestion would be:
```python
def get_sensors_by_device_class(
    _er: EntityRegistry,
    _hass: HomeAssistant,
    device_class: str,
    include_all: bool = False,
) -> list:
    """Get sensors of required class from entity registry."""

    result = [
        state.entity_id
        for state in filter(
            lambda state:
                "device_class" in state.attributes
                and state.domain == Platform.SENSOR
                and state.attributes["device_class"] == SensorDeviceClass.TEMPERATURE,
            _hass.states.async_all(),
        )
    ]

    result.sort()

    if include_all:
        all_sans_result = list({state.entity_id for state in _hass.states.async_all()} - set(result))
        all_sans_result.sort()
        result += all_sans_result

    result = filter(
        lambda entity_id: all(sensor_type.to_shortform() not in entity_id for sensor_type in SensorType),
        result,
    )

    return result
```

1. Get all states with correct device class and domain. No reason to stick to the entity registry here.
2. If advanced mode add all other entities sorted at the end of the list.
3. Always remove entity ids containing our SensorType

This way we have the obvious choices sorted at the top of the list and the rest sorted bellow. Should be easy to navigate.

I am a bit unsure about step 3 since we don't know if someone would want to feed back our sensors in the platform itself but currently I don't see a reason to do so and it really cleans up the list.

What do you think?

**Comment by IATkachenko** [comment]
Looks promising.
I personally don't like `lambda` calls: it hard to test and read. 

Also we should check original_domain, like `async_get_device_class_lookup` do.
And I'm going to check `unit_of_measurements`, for correct sensor detection.

I'm agree with you abut excluding own sensors from list -- it is a good idea. It not looks like this data should be used in `thermal_comfort`

Will update PR today, may be on holidays.

**Comment by rautesamtr** [comment]
> I personally don't like lambda calls: it hard to test and read.

Understandable. Here you go:
``` python
from homeassistant.core import …, State, …
…
def get_sensors_by_device_class(
    _er: EntityRegistry,
    _hass: HomeAssistant,
    device_class: str,
    include_all: bool = False,
) -> list:
    """Get sensors of required class from entity registry."""

    def filter_for_device_class_sensor(state: State) -> bool:
        """Filter states by Platform.SENSOR and SensorDeviceClass.TEMPERATURE."""
        if (
            "device_class" in state.attributes
            and state.domain == Platform.SENSOR
            and state.attributes["device_class"] == device_class
        ):
            return True

    def filter_thermal_comfort_ids(entity_id: str) -> bool:
        """Filter out device_ids containing our SensorType."""
        return all(
            sensor_type.to_shortform() not in entity_id for sensor_type in SensorType
        )

    result = [
        state.entity_id
        for state in filter(
            filter_for_device_class_sensor,
            _hass.states.async_all(),
        )
    ]

    result.sort()

    if include_all:
        all_sans_result = list(
            {state.entity_id for state in _hass.states.async_all()} - set(result)
        )
        all_sans_result.sort()
        result += all_sans_result

    result = filter(
        filter_thermal_comfort_ids,
        result,
    )

    return result
```

**Comment by IATkachenko** [comment]
@rautesamtr, why you suggesting exclude only discovered sensors
```python
{state.entity_id for state in _hass.states.async_all()} - set(result)
```
? Pressure sensors, for example, will be added to list. Like an automatizations, binary sensors etc. Even advanced uses should not use automatization as source of temperature for the integration.

What will we miss if we are using `set(_er.entities)` here?

**Comment by rautesamtr** [comment]
Like i said this excludes to much -> helpers

**Comment by rautesamtr** [comment]
Also people will open issues when their input entities isn't showing up, so we possibly have to start creating workarounds for specific cases. Again possible more maintenance for restricting advanced users.

**Comment by IATkachenko** [comment]
Helpers, like command_line sensors, looks useful.

Then we should use opposite way: create close list for exclude. Like automatization, scene, binary sensor, media, etc. This object should not be used as data source in any case. How do you like this idea?

**Comment by rautesamtr** [comment]
Helpers like https://www.home-assistant.io/integrations/input_number/ they are called helpers in the frontend. Sorry for not being explicit enough.

> Then we should use opposite way: create close list for exclude. Like automatization, scene, binary sensor, media, etc. This object should not be used as data source in any case. How do you like this idea?

Great idea. So to sum things up:
 * Exclude sensors that have their device_class set to one that we aren't expecting (e.g. exclude pressure sensors)
 * Exclude domains we know doesn't provide any useful input
 * Keep stuff that can give us a number as input
 * Keep sensors without device_class set (most likely that haven't implemented that feature yet so may be valid input)
 * Exclude entities ids with our SensorTypes in their id.

Did i miss anything?

**Comment by IATkachenko** [comment]
Looks good enough.

Not sure about "keep" part: we are going to exclude from `_hass.states.async_all()` based on set of predefined criteria. It should be taken into account while criteria formation, not while filtering part.
For now I see criteria like:
 1. devicec

**Comment by IATkachenko** [comment]
Looks like crazy beast, but I got very good results with this filters.

File: custom_components/thermal_comfort/config_flow.py [hunk 7]
```diff
@@ -21,6 +23,287 @@
...
+        SensorDeviceClass.PM25,
+        SensorDeviceClass.POWER_FACTOR,
+        SensorDeviceClass.POWER,
+        SensorDeviceClass.PRESSURE,
+        SensorDeviceClass.SIGNAL_STRENGTH,
+        SensorDeviceClass.SULPHUR_DIOXIDE,
+        SensorDeviceClass.TIMESTAMP,
+        SensorDeviceClass.VOLATILE_ORGANIC_COMPOUNDS,
+        SensorDeviceClass.VOLTAGE,
+    ]
```
Comment (rautesamtr) [inline custom_components/thermal_comfort/config_flow.py:61]: We could shrink this down if you want
```python
    DEVICE_CLASS_FOR_EXCLUDE = list(SensorDeviceClass)
    DEVICE_CLASS_FOR_EXCLUDE.remove(SensorDeviceClass.TEMPERATURE)
    DEVICE_CLASS_FOR_EXCLUDE.remove(SensorDeviceClass.HUMIDITY)
```

File: custom_components/thermal_comfort/config_flow.py [hunk 8]
```diff
@@ -21,6 +23,287 @@
...
+        _state: State, _list: list[SensorDeviceClass], should_be_in: bool = True
+    ) -> bool:
+        collected_device_class = _state.attributes.get(
+            "device_class", _state.attributes.get("original_device_class")
+        )
+        # XNOR
+        return not ((collected_device_class in _list) ^ should_be_in)
+
+    def filter_for_device_class_sensor(state: State) -> bool:
+        """Filter states by Platform.SENSOR and SensorDeviceClass.TEMPERATURE."""
```
Comment (rautesamtr) [inline custom_components/thermal_comfort/config_flow.py:106]: and SensorDeviceClass.HUMIDITY

**Review by rautesamtr — APPROVED** [review]
Wow. Works great. I think we can merge this.

File: custom_components/thermal_comfort/config_flow.py (hunk 7 above)
Comment (IATkachenko) [reply custom_components/thermal_comfort/config_flow.py:61]: Actually it is not what we really want. If some useful for us device class will be added to `SensorDeviceClass`, then we will miss it.

**Comment by IATkachenko** [comment]
> Wow. Works great. I think we can merge this.

It could be merged only after squashing/rebasing...

**Comment by rautesamtr** [comment]
Thanks for polishing and getting this feature in before 1.4.0. I think this provides a huge UX boost.

**Comment by IATkachenko** [comment]
Welcome!

**Pull request merged**